subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginCheckInService.java

```diff
@@ -0,0 +1,95 @@
+    private final boolean isConfigurationCacheEnabled;
+
+    public DefaultGradleEnterprisePluginCheckInService(
+        BuildModelParameters buildModelParameters,
+        GradleEnterprisePluginManager manager,
+        DefaultGradleEnterprisePluginAdapter adapter
+    ) {
+        this.manager = manager;
+        this.adapter = adapter;
+        this.isConfigurationCacheEnabled = buildModelParameters.isConfigurationCache();
+    }
+
+    // Used just for testing
+    public static final String UNSUPPORTED_TOGGLE = "org.gradle.internal.unsupported-enterprise-plugin";
+    public static final String UNSUPPORTED_TOGGLE_MESSAGE = "Enterprise plugin unsupported due to secret toggle";
+
+    // For Gradle versions 8+, configuration caching builds are not compatible with Gradle Enterprise plugin < 3.12
+    public static final VersionNumber MINIMUM_SUPPORTED_PLUGIN_VERSION_FOR_CONFIGURATION_CACHING = VersionNumber.version(3, 12);
+    public static final String UNSUPPORTED_PLUGIN_DUE_TO_CONFIGURATION_CACHING_MESSAGE = String.format("The Gradle Enterprise plugin has been disabled as it is " +
+            "incompatible with this version of Gradle and the configuration caching feature - please upgrade to version %s.%s or later of the Gradle Enterprise plugin to restore functionality.",
+        MINIMUM_SUPPORTED_PLUGIN_VERSION_FOR_CONFIGURATION_CACHING.getMajor(),
+        MINIMUM_SUPPORTED_PLUGIN_VERSION_FOR_CONFIGURATION_CACHING.getMinor());
+
+    @Override
+    public GradleEnterprisePluginCheckInResult checkIn(GradleEnterprisePluginMetadata pluginMetadata, GradleEnterprisePluginServiceFactory serviceFactory) {
+        if (Boolean.getBoolean(UNSUPPORTED_TOGGLE)) {
+            manager.unsupported();
+            return checkInResult(UNSUPPORTED_TOGGLE_MESSAGE, () -> {
+                throw new IllegalStateException();
+            });
+        }
+        if (isConfigurationCacheEnabled && isUnsupported(pluginMetadata.getVersion())) {
+            manager.unsupported();
+            return checkInResult(UNSUPPORTED_PLUGIN_DUE_TO_CONFIGURATION_CACHING_MESSAGE, () -> {
+                throw new IllegalStateException();
+            });
+        }
+        GradleEnterprisePluginServiceRef ref = adapter.register(serviceFactory);
+        manager.registerAdapter(adapter);
+        return checkInResult(null, () -> ref);
+    }
+
+    private static GradleEnterprisePluginCheckInResult checkInResult(String unsupportedMessage, Supplier<GradleEnterprisePluginServiceRef> pluginServiceRefSupplier) {
+        return new GradleEnterprisePluginCheckInResult() {
+            @Override
+            public String getUnsupportedMessage() {
+                return unsupportedMessage;
+            }
+
+            @Override
+            public GradleEnterprisePluginServiceRef getPluginServiceRef() {
+                return pluginServiceRefSupplier.get();
+            }
+        };
+    }
+
+    private static boolean isUnsupported(String pluginVersion) {
```

Comment:
This should be called something like `isUnsupportedWithConfigurationCaching()` or all of the checks should move in here.

Reply:
Done in https://github.com/gradle/gradle/commit/0bcb7ddc86788fd6d0e45f6a99bc2c19161d893b